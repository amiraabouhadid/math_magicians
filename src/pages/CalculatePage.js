import React from 'react';
import Calculator from '../components/Calculator';

const CalculatePage = () => (
  <div className=" container-fluid">
    <div className="row p-4 m-5">
      <h3 className="col-4"> Let &apos;s do some math!</h3>
      <div className="col-8">
        <Calculator />
      </div>
    </div>

  </div>
);
export default CalculatePage;
