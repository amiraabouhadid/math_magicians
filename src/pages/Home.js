import React from 'react';
import { LoremIpsum } from 'react-lorem-ipsum';

const Home = () => (
  <div className="p-5">
    <h2>Welcome to our page! </h2>

    <LoremIpsum />
    <LoremIpsum />
  </div>
);
export default Home;
