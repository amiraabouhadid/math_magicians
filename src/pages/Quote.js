import React from 'react';

class Quote extends React.Component {
  constructor(props) {
    super(props);
    this.state = {
      quote: '',
      author: '',
    };
  }

  async componentDidMount() {
    await fetch('https://random-math-quote-api.herokuapp.com/', {
      contentType: 'application/json',
    })
      .then((response) => response.json())
      .then((response) => {
        this.setState({
          quote: response.quote,
          author: response.author,
        });
      })
      .catch((err) => console.error(err));
  }

  render() {
    const { quote, author } = this.state;
    return (
      <div className="m-5 p-5 ">
        <h4 className="p-5">
          {quote}
          <span className="mx-3">â€”</span>
          {author}
        </h4>

      </div>
    );
  }
}

export default Quote;
