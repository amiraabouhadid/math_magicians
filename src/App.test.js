import { render, screen } from '@testing-library/react';
import renderer from 'react-test-renderer';
import operate from './logic/operate';
import calculate from './logic/calculate';
import Home from './pages/Home';
import CalculatePage from './pages/CalculatePage';
import Quote from './pages/Quote';

describe('operate mathematical functions', () => {
  test('adds 1 + 2 to equal 3', () => {
    expect(operate(1, 2, '+')).toBe('3');
  });
  test('adds 2 - 1 to equal 1', () => {
    expect(operate(2, 1, '-')).toBe('1');
  });
  test('adds 2 / 1 to equal 2', () => {
    expect(operate(2, 1, 'รท')).toBe('2');
  });
  test('adds 2 / 1 to equal 2', () => {
    expect(operate(2, 1, 'รท')).toBe('2');
  });
  test('adds 2 x 1 to equal 2', () => {
    expect(operate(2, 1, 'x')).toBe('2');
  });
  test("adds 2 / 0 to return can't divide by zero", () => {
    expect(operate(2, 0, 'รท')).toBe("Can't divide by 0.");
  });
  test('returns modulo of 1%2', () => {
    expect(operate(1, 2, '%')).toBe('1');
  });
  test("returns can't find module as can't divide by 0", () => {
    expect(operate(1, 0, '%')).toBe("Can't find modulo as can't divide by 0.");
  });
});
describe('calculate logic ', () => {
  test('AC button', () => {
    expect(calculate({ total: 0, next: null, operation: null }, 'AC')).toStrictEqual({ total: 0, next: null, operation: null });
  });
  test('= button in case of operation and next value', () => {
    expect(calculate({ total: 0, next: '3', operation: '+' }, '=')).toStrictEqual({ total: '3', next: null, operation: null });
  });
  test('= button in case of no operation', () => {
    expect(calculate({ total: 0, next: null, operation: null }, '=')).toStrictEqual({ });
  });
  test('+/- button in case of next value', () => {
    expect(calculate({ total: 0, next: '3', operation: null }, '+/-')).toStrictEqual({ total: 0, next: '-3', operation: null });
  });
  test('+/- button in case of total ', () => {
    expect(calculate({ total: '1', next: null, operation: null }, '+/-')).toStrictEqual({ total: '-1', next: null, operation: null });
  });
  test('+/- button in case of no total or next value', () => {
    expect(calculate({ total: 0, next: null, operation: null }, '+/-')).toStrictEqual({ });
  });
  test('Number button in case of 0 and next is 0 ', () => {
    expect(calculate({ total: '0', next: '0', operation: null }, '0')).toStrictEqual({ });
  });
  test('Number button in case of number with existing next value and operation ', () => {
    expect(calculate({ total: 0, next: '3', operation: '+' }, '3')).toStrictEqual({ total: 0, next: '33', operation: '+' });
  });
  test('Number button in case of number with existing operation but no next value', () => {
    expect(calculate({ total: 0, next: null, operation: '+' }, '3')).toStrictEqual({ total: 0, next: '3', operation: '+' });
  });
  test('Number button in case of number with existing next value and no operation', () => {
    expect(calculate({ total: 0, next: '3', operation: null }, '3')).toStrictEqual({ total: 0, next: '33' });
  });
  test('Number button in case of number with no  next value and no operation', () => {
    expect(calculate({ total: 0, next: null, operation: null }, '3')).toStrictEqual({ total: 0, next: '3' });
  });
  test('. button when no total, next or operation is given', () => {
    expect(calculate({ total: 0, next: null, operation: null }, '.')).toStrictEqual({ total: '0.' });
  });
  test('. button when only next  is given and does not include .', () => {
    expect(calculate({ total: 0, next: '3', operation: null }, '.')).toStrictEqual({ total: 0, next: '3.', operation: null });
  });
  test('. button when only next  is given and does include .', () => {
    expect(calculate({ total: 0, next: '3.', operation: null }, '.')).toStrictEqual({ total: 0, next: '3.', operation: null });
  });
  test('. button when only operation is given', () => {
    expect(calculate({ total: 0, next: null, operation: '+' }, '.')).toStrictEqual({ next: '0.' });
  });
  test('. button when only total is given and does not include .', () => {
    expect(calculate({ total: '3', next: null, operation: null }, '.')).toStrictEqual({ total: '3.' });
  });
  test('. button when only total is given and includes .', () => {
    expect(calculate({ total: '3.', next: null, operation: null }, '.')).toStrictEqual({ });
  });
  test("User pressed an operation after pressing '='", () => {
    expect(calculate({ total: '3.', next: null, operation: null }, '+')).toStrictEqual({ total: '3.', next: null, operation: '+' });
  });
  test('User pressed an operation button and there is an existing operation and no next value', () => {
    expect(calculate({ total: '3.', next: null, operation: '+' }, '-')).toStrictEqual({ total: '3.', next: null, operation: '-' });
  });
  test('User pressed an operation button and there is an existing operation and a next value', () => {
    expect(calculate({ total: '3', next: '1', operation: '+' }, '-')).toStrictEqual({ total: '4', next: null, operation: '-' });
  });
  test("The user hasn't typed a number yet and no next value, just save the operation", () => {
    expect(calculate({ total: 0, next: null, operation: null }, '-')).toStrictEqual({ operation: '-' });
  });
  test("The user hasn't typed a number yet but with next value, just save the operation", () => {
    expect(calculate({ total: 0, next: '3', operation: null }, '-')).toStrictEqual({ total: '3', next: null, operation: '-' });
  });
});
describe('home page component', () => {
  test('renders header text', () => {
    render(<Home />);
    const headerElement = screen.getByText(/Welcome to our page!/i);
    expect(headerElement).toBeInTheDocument();
  });
  test('renders correctly', () => {
    const home = renderer
      .create(<Home />)
      .toJSON();
    expect(home).toMatchSnapshot();
  });
});
describe('quote page component', () => {
  test('renders correctly', () => {
    const quote = renderer
      .create(<Quote />)
      .toJSON();
    expect(quote).toMatchSnapshot();
  });
});
describe('calculator page component', () => {
  test('renders correctly', () => {
    const calculatePage = renderer
      .create(<CalculatePage />)
      .toJSON();
    expect(calculatePage).toMatchSnapshot();
  });
  test('renders header text', () => {
    render(<CalculatePage />);
    const header3Element = screen.getByText(/Lets do some math!/i);
    expect(header3Element).toBeInTheDocument();
  });
  test('renders AC button', () => {
    render(<CalculatePage />);
    const ACButton = screen.getByText(/AC/i);
    expect(ACButton).toBeInTheDocument();
  });
  test('renders numbers buttons and total set at 0', () => {
    render(<CalculatePage />);
    const NumbersButtons = screen.getAllByText(/[0-9]/i);
    expect(NumbersButtons).toHaveLength(11);
  });
  test('renders equal button', () => {
    render(<CalculatePage />);
    const EqualButton = screen.getByText('=');
    expect(EqualButton).toBeInTheDocument();
  });
  test('renders operations buttons', () => {
    render(<CalculatePage />);
    const OperationsButtons = screen.getAllByText(/[(x)(+)(รท)(-)]{1}/i);
    expect(OperationsButtons).toHaveLength(4);
  });
  test('renders +/- button', () => {
    render(<CalculatePage />);
    const ChangeSignButton = screen.getByText('+/-');
    expect(ChangeSignButton).toBeInTheDocument();
  });
});
