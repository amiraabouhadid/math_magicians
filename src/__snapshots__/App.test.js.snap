// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Navlink tests Quote Link 1`] = `
<div
  className="m-5 p-5 "
>
  <h4
    className="p-5"
  >
    
    <span
      className="mx-3"
    >
      —
    </span>
    
  </h4>
</div>
`;

exports[`calculator page component renders correctly 1`] = `
<div
  className=" container-fluid"
>
  <div
    className="row p-4 m-5"
  >
    <h3
      className="col-4"
    >
      Lets do some math!
    </h3>
    <div
      className="col-8"
    >
      <div
        className="container-fluid"
      >
        <div
          className="py-2 container"
        >
          <table
            className="table table-bordered"
          >
            <tbody>
              <tr>
                <td
                  className="bg-secondary p-4 text-end text-white w-100"
                  colSpan="4"
                >
                  <p>
                    0
                  </p>
                </td>
              </tr>
              <tr>
                <td
                  className="bg-light"
                  colSpan="1"
                >
                  <button
                    className="w-100 btn btn-light"
                    disabled={false}
                    onClick={[Function]}
                    type="button"
                    value="AC"
                  >
                    AC
                  </button>
                </td>
                <td
                  className="bg-light"
                  colSpan="1"
                >
                  <button
                    className="w-100 btn btn-light"
                    disabled={false}
                    onClick={[Function]}
                    type="button"
                    value="+/-"
                  >
                    +/-
                  </button>
                </td>
                <td
                  className="bg-light"
                  colSpan="1"
                >
                  <button
                    className="w-100 btn btn-light"
                    disabled={false}
                    onClick={[Function]}
                    type="button"
                    value="%"
                  >
                    %
                  </button>
                </td>
                <td
                  className="bg-warning"
                  colSpan="1"
                >
                  <button
                    className="w-100 btn btn-warning"
                    disabled={false}
                    onClick={[Function]}
                    type="button"
                    value="÷"
                  >
                    ÷
                  </button>
                </td>
              </tr>
              <tr>
                <td
                  className="bg-light"
                  colSpan="1"
                >
                  <button
                    className="w-100 btn btn-light"
                    disabled={false}
                    onClick={[Function]}
                    type="button"
                    value="7"
                  >
                    7
                  </button>
                </td>
                <td
                  className="bg-light"
                  colSpan="1"
                >
                  <button
                    className="w-100 btn btn-light"
                    disabled={false}
                    onClick={[Function]}
                    type="button"
                    value="8"
                  >
                    8
                  </button>
                </td>
                <td
                  className="bg-light"
                  colSpan="1"
                >
                  <button
                    className="w-100 btn btn-light"
                    disabled={false}
                    onClick={[Function]}
                    type="button"
                    value="9"
                  >
                    9
                  </button>
                </td>
                <td
                  className="bg-warning"
                  colSpan="1"
                >
                  <button
                    className="w-100 btn btn-warning"
                    disabled={false}
                    onClick={[Function]}
                    type="button"
                    value="x"
                  >
                    x
                  </button>
                </td>
              </tr>
              <tr>
                <td
                  className="bg-light"
                  colSpan="1"
                >
                  <button
                    className="w-100 btn btn-light"
                    disabled={false}
                    onClick={[Function]}
                    type="button"
                    value="4"
                  >
                    4
                  </button>
                </td>
                <td
                  className="bg-light"
                  colSpan="1"
                >
                  <button
                    className="w-100 btn btn-light"
                    disabled={false}
                    onClick={[Function]}
                    type="button"
                    value="5"
                  >
                    5
                  </button>
                </td>
                <td
                  className="bg-light"
                  colSpan="1"
                >
                  <button
                    className="w-100 btn btn-light"
                    disabled={false}
                    onClick={[Function]}
                    type="button"
                    value="6"
                  >
                    6
                  </button>
                </td>
                <td
                  className="bg-warning"
                  colSpan="1"
                >
                  <button
                    className="w-100 btn btn-warning"
                    disabled={false}
                    onClick={[Function]}
                    type="button"
                    value="-"
                  >
                    -
                  </button>
                </td>
              </tr>
              <tr>
                <td
                  className="bg-light"
                  colSpan="1"
                >
                  <button
                    className="w-100 btn btn-light"
                    disabled={false}
                    onClick={[Function]}
                    type="button"
                    value="1"
                  >
                    1
                  </button>
                </td>
                <td
                  className="bg-light"
                  colSpan="1"
                >
                  <button
                    className="w-100 btn btn-light"
                    disabled={false}
                    onClick={[Function]}
                    type="button"
                    value="2"
                  >
                    2
                  </button>
                </td>
                <td
                  className="bg-light"
                  colSpan="1"
                >
                  <button
                    className="w-100 btn btn-light"
                    disabled={false}
                    onClick={[Function]}
                    type="button"
                    value="3"
                  >
                    3
                  </button>
                </td>
                <td
                  className="bg-warning"
                  colSpan="1"
                >
                  <button
                    className="w-100 btn btn-warning"
                    disabled={false}
                    onClick={[Function]}
                    type="button"
                    value="+"
                  >
                    +
                  </button>
                </td>
              </tr>
              <tr>
                <td
                  className="bg-light"
                  colSpan="2"
                >
                  <button
                    className="w-100 btn btn-light"
                    disabled={false}
                    onClick={[Function]}
                    type="button"
                    value="0"
                  >
                    0
                  </button>
                </td>
                <td
                  className="bg-light"
                  colSpan="1"
                >
                  <button
                    className="w-100 btn btn-light"
                    disabled={false}
                    onClick={[Function]}
                    type="button"
                    value="."
                  >
                    .
                  </button>
                </td>
                <td
                  className="bg-warning"
                  colSpan="1"
                >
                  <button
                    className="w-100 btn btn-warning"
                    disabled={false}
                    onClick={[Function]}
                    type="button"
                    value="="
                  >
                    =
                  </button>
                </td>
              </tr>
            </tbody>
          </table>
        </div>
      </div>
    </div>
  </div>
</div>
`;

exports[`home page component renders correctly 1`] = `
<div
  className="p-5"
>
  <h2>
    Welcome to our page!
  </h2>
  <p>
    Lorem ipsum odor amet, consectetuer adipiscing elit. Natoque pulvinar fringilla vehicula vitae rhoncus congue phasellus pretium. Leo a ullamcorper nam ultrices maximus mattis. At est molestie elit quis pellentesque facilisi? Viverra a ornare morbi ut taciti; nisi semper. Rutrum porttitor nam laoreet non; quis quis primis vulputate porttitor. Dictumst fermentum odio nisi tempus, nisi sagittis litora. Nunc integer dui sem interdum porttitor mollis.
  </p>
  <p>
    Lorem ipsum odor amet, consectetuer adipiscing elit. Ultrices consectetur nascetur maximus condimentum; vivamus proin. Hac blandit sodales suspendisse adipiscing aliquam quisque morbi felis natoque. Aliquet non odio conubia convallis quisque. Urna ipsum porttitor imperdiet fringilla tincidunt; pharetra efficitur fermentum augue. Fringilla erat pretium donec rutrum duis eros erat. Turpis venenatis pharetra leo parturient aliquam lobortis felis. Ex primis sapien lacus faucibus est pretium lectus porta. Rutrum velit vitae consequat, nisl donec fusce ante morbi. Magna tempor non magnis; consectetur felis orci proin neque.
  </p>
</div>
`;

exports[`quote page component renders correctly 1`] = `
<div
  className="m-5 p-5 "
>
  <h4
    className="p-5"
  >
    
    <span
      className="mx-3"
    >
      —
    </span>
    
  </h4>
</div>
`;
